### DISCUSSION TITLE
Simple way to have Stores in Alpine

### DISCUSSION THREAD

**Discussion by ahmedkandel** [alpinejs/alpine#497] [4. General]
Hello,

I was working on a project using Alpine where sharing state between components was necessary.
This could be achieved by dispatching events from one component using `$dispatch('custom-event', { foo: 'bar' })` , Then listen to the event after it bubbles up using `x-on:custom-event.document` on the other component where the attached state can be retrieved from `$event.detail`.

But this is not the level of reactivity we were seeking so the obvious solution is to use a store like Vuex and Redux does. I have found Spruce by @ryangjchandler to be interesting but I was thinking of a simple snippet with no more libraries to include in the project, So I would like to share my snippet with you maybe someone finds it useful.

As a final result, this Alpine.JS $store is supporting read-only, writable, and derived properties also sharing methods and watcher for properties mutations.

**Here is a codepen to demo this store in action https://codepen.io/tomorrowsystem/pen/bGVmozb?editors=1011**

### Setup:

First, in a script tag define the state you would like to share between components, it may contain primitives, objects, and functions:
```javascript
<script>
    // User defined store data with default values
    const store = {
        key1: 0,
        key2: 0,
        key3: {
            foo: 'bar',
        },
        key4: ['foo', 'bar'],
        key5: 'readonly',
        key6: 0,
        func1() {
            this.key1++;
        },
        func2() {
            this.key2++;
        },
        derived() {
            return this.key4.join(', ').toUpperCase();
        }
    };
</script>
```
Then add the snippet for the store logic in a script tag:
```javascript
<script>
        window.$store = new Proxy({
        ...store,
        subscriptions: [], subscribe(props, component) {
            props.forEach(prop => this.subscriptions[prop] = [...(this.subscriptions[prop] || []), component]); //subscribe
        },
        watchers: [], watch(props, callback) {
            props.forEach(prop => this.watchers[prop] = [...(this.watchers[prop] || []), callback]); //watch
        },
        protected: new Set(), readonly(props) {
            props.forEach(prop => this.protected.add(prop)); //make readonly
        },
    },
    {
        set(obj, prop, value) {
            (obj.watchers[prop] || []).forEach(callback => callback(obj[prop], value)); //callback watcher
            (obj.subscriptions[prop] || []).forEach(component => component.__x.$data.$NaN = NaN); //notify component
            if (obj.protected.has(prop)) return false; //reject readonly
            return Reflect.set(...arguments); //set a property on object
        }
    });
</script>
```
Done!

### Usage:

First, subscribe to the store properties you need to be reactive in you component like this:
```
x-init="$store.subscribe(['key1','key2'], $el)"
```

Done, now whenever those properties get updated you component will update and get the new values, e.g. you can use x-text direct to show the updated value:
```
x-text="$store.key1"
```

You may also need to update the store property value or bind it to an input:
```
x-model.number="$store.key1"
```

Call functions that affect the state from any component and it will reflect to all subscriber components:
```
x-on:click="$store.func1()"
```

Functions can return derived or computed properties:
```
x-text="$store.derived()"
```

### Extra:

You can make a store property read-only so it doesn't get accidentally mutated by components as follows:
```
x-init="$store.readonly(['key5']); $store.subscribe(['key5'], $el);"
```

Also, you may like to have watchers to invoke a callback whenever this state changes:
```
x-init="$store.watch(['key6'], (oldVal, newVal)=> console.log(oldVal, newVal))"
```

If you find this interesting please let me know so I explain it more in detail and how it is working. Please also feel free to suggest any modifications or a better approach.

Thanks.

**Comment by ryangjchandler** [comment]
All of that behaviour is already provided by Spruce in an almost identical way. I'm biased but it seems like reinventing the wheel for no reason, the JS you've written is probably comparable in size to Spruce too. 

The only thing Spruce doesn't have is readonly properties, but I can't think of many use cases where these would be needed.

**Reply by ryangjchandler** [reply]
Does your code handle nested properties?

**Reply by ahmedkandel** [reply]
Yes in the codepen there is an example for a nested object using assignment operators like the ways SvelteJS is doing it.

This is not a comparison with Spruce, it is just a simpler way to do stores.

**Comment by HugoDF** [comment]
I believe the plan for Alpine v3 is to have `store` in core since we can model a `store` entity to be a component without a template/rendering, just the reactive data.

I like the use of `NaN` since it's not equal to itself in `component.__x.$data.$NaN = NaN`, although a similar effect would be achieved by setting any property to a new object/array or using one of the Alpine.js methods: `__x.updateElements()`

**Reply by ryangjchandler** [reply]
Yeah, relying on the `updateElements()` method is much better than updating the data directly in my opinion, hence why I switched to that with Spruce.

**Reply by ryangjchandler** [reply]
And as @HugoDF mentioned, there are plans for a core global state layer for v3, although it's not set in stone.

**Reply by ahmedkandel** [reply]
Yep, I like to use `__x.updateElements()` while for sure will be better if Alpine exposes an official method to force update the component.

Why I prefer `__x.$data` maybe because Alpine is currently been refactored frequently and the method names may change which the data object has fewer chances to change.

**Reply by ryangjchandler** [reply]
I think any changes in the method names would be considered a breaking change since it's semi exposed method.

**Comment by ahmedkandel** [comment]
> Alpine.js offers you the reactive and declarative nature of big frameworks like Vue or React at a much lower cost

Alpine aims to simplify things not reinventing the wheel, The same thing here I want to make stores more simple with no extra libraries or dependencies. Also performance wise and size this snippet updates only components that subscribe to certain keys, not all subscribers. With a minimal footprint and some extra like read-only properties that I see may be useful in many ways.

**Reply by ryangjchandler** [reply]
Spruce is still pre v1, if you feel there's missing features open an issue and I'd happily take a look.

**Reply by ahmedkandel** [reply]
Thanks @ryangjchandler I like Spruce and would like to contribute but the main reason I like Alpine is to avoid build steps and extra dependencies/libraries so I prefer to sprinkle JS over using external lib.

I will be so much happy when Alpine has its own store management builtin.

**Reply by ryangjchandler** [reply]
Spruce has no build steps it can be used through a CDN the same way as Alpine, and it has the same API as yours in the sense you subscribe to a store.

**Reply by ryangjchandler** [reply]
Spruce has also been designed to be as close to the internally discussed APIs for integrated global stores.

**Comment by ahmedkandel** [comment]
Guys @HugoDF @ryangjchandler @SimoTod can you suggest a better way for store subscriptions?

**Reply by ryangjchandler** [reply]
With the current open API, there's no easy way to integrate with the component's available data without hacking it with x-init, as Spruce does it.

**Comment by kwemou** [comment]
Dear all,
Please do not add too much functionality to Alpinejs.
Adding a store in Alpinejs just make it looses it minimalism. 
It is better to make sure that it integrates well with existing redux,  flux or whatever libs to provides Stores functionality.
Thank you and with best regards 
Emeric
